### PR TITLE
Basic Conway's Functionalities

### PR DESCRIPTION
**Pull request by JereSalo**
## Summary
Implementation of basic **Conway's Game of Life** functionalities. The main character in this PR is the *Grid* data type.

## Changes
- Create a *Grid* of cells in different ways
     1. Empty (with Dead cells)
     2. Random
     3. Custom
- Get the next generation of a Grid's cells

### CODE REVIEW

**Comment by JereSalo** [comment]
I realized I didn't fill the readme file. I will do that later

**Comment by iinaki** [comment]
I notice that some commit messages don't follow the imperative mood that is recommended [here](https://cbea.ms/git-commit/), I would tell you to think whenever you are writing a commit to try to complete the frase "If applied, this commit will `your subject line here`". You don't have to change anything, just to keep it in mind for the future.

**Comment by iinaki** [comment]
Another thing I noticed is that you run `cargo init` in the main branch, you should never push to main without a PR, even for that small things!

**Review by iinaki — CHANGES_REQUESTED** [review]
This is important to me, try to run `cargo fmt` and  `cargo clippy` ( I use `cargo clippy --all-targets --all-features -- -D warnings` ) before wanting to merge to main.

File: src/grid.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        count
+    }
+
+    // Get all neighbors of a given position (not the actual Cell velues, but their positions in the grid)
+    // Handles the case in which there might be a position with fewer neighbors. Like [0,0]
+    pub fn get_neighbors(&self, x:usize, y:usize) -> Vec<(usize, usize)>{
+        let mut neighbors: Vec<(usize, usize)> = Vec::new();
+
+        // Relative positions
+        let deltas = [
```
Comment (iinaki) [inline src/grid.rs]: Here I believe `deltas` can be a defined constant.

File: src/grid.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            (1, -1), (1, 0), (1, 1),
+        ];
+
+        for (dx, dy) in &deltas {
+            // I use isize because result can be negative.
+            let new_x = x as isize + dx;
+            let new_y = y as isize + dy;
+
+            // Only push neighbor to neighbors if it's position is valid
+            if new_x >= 0 && new_x < self.rows() as isize &&
```
Comment (iinaki) [inline src/grid.rs]: I would move this check `new_x >= 0 && new_x < self.rows() as isize && new_y >= 0 && new_y < self.cols() as isize` to be an abstraction, for example something called `position_is_out_of_bounds` .

File: src/grid.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    // For each cell in the grid calculates if in the next gen is going to be alive or not. Returns the next generation.
+    pub fn calculate_next_gen(&self) -> Vec<Vec<Cell>> {
+        let mut next_gen: Vec<Vec<Cell>> = vec![vec![Cell::Dead; self.cols()]; self.rows()];
+
+        for i in 0..self.rows(){
+            for j in 0..self.cols(){
+                let current_gen_cell = self.cells[i][j];
+                let alive_neighbors = self.count_alive_neighbors(i, j);
```
Comment (iinaki) [inline src/grid.rs:95]: Something a little bit bigger that I see is that when you check for the neighbours positions you iterate twice the neighbours of each cell, first to store their positions and then to count if they are dead or alive. You can change a bit the logic of how `get_neighbors` and `count_alive_neighbors` work to iterate only once the neighbours. Also you can avoid storing them in a `Vec<(usize, usize)>` as you don't really need this!

**Review by iinaki — CHANGES_REQUESTED** [review]
I left some comments, they are some minor things that you can change and then I'll approve! I liked how you handled the calculation of the next gen, and I'm exited to see what you come up with when you implement te UI!

File: src/grid.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        Grid::new_from_cells(vec![
+            vec![Cell::Dead, Cell::Alive, Cell::Dead, Cell::Dead],
+            vec![Cell::Alive, Cell::Alive, Cell::Dead, Cell::Dead],
+            vec![Cell::Dead, Cell::Dead, Cell::Alive, Cell::Dead],
+            vec![Cell::Dead, Cell::Dead, Cell::Dead, Cell::Alive],
+        ])
+    }
+
+    #[test]
+    fn test_grid_initialization() {
```
Comment (iinaki) [inline src/grid.rs]: The `test_` prefix isn't needed, I think that with the #[test] attribute is enough.

**Review by iinaki — CHANGES_REQUESTED** [review]

**Comment by JereSalo** [comment]
> I notice that some commit messages don't follow the imperative mood that is recommended [here](https://cbea.ms/git-commit/), I would tell you to think whenever you are writing a commit to try to complete the frase "If applied, this commit will `your subject line here`". You don't have to change anything, just to keep it in mind for the future.

Thanks! I will take it into account.

File: src/grid.rs (hunk 1 above)
Comment (JereSalo) [reply src/grid.rs]: Done!

File: src/grid.rs (hunk 2 above)
Comment (JereSalo) [reply src/grid.rs]: Done! I implemented it as a private method in Grid. Do you think that's ok?

File: src/grid.rs (hunk 3 above)
Comment (JereSalo) [reply src/grid.rs:95]: You are right! Now the code is way more concise. Thanks!

File: src/grid.rs (hunk 4 above)
Comment (JereSalo) [reply src/grid.rs]: Done!

File: src/grid.rs (hunk 2 above)
Comment (iinaki) [reply src/grid.rs]: Yeah thats great!

**Comment by iinaki** [comment]
Everything looks good to me! But I ran `cargo fmt` and `cargo clippy --all-targets --all-features -- -D warnings` and still got some warnings, fix that and I'll approve!

**Review by iinaki — APPROVED** [review]

**Pull request merged**